src/objective-c/ProtoRPC/ProtoRPC.m

```diff
@@ -23,9 +23,166 @@
+                                                   responseHandler:handler
+                                                       callOptions:callOptions
+                                                     responseClass:responseClass];
+    [_call writeWithMessage:message];
+    [_call finish];
+  }
+  return self;
+}
+
+- (void)cancel {
+  [_call cancel];
+  _call = nil;
+}
+
+@end
+
+@interface GRPCStreamingProtoCall ()<GRPCResponseHandler>
+
+@end
+
+@implementation GRPCStreamingProtoCall {
+  GRPCRequestOptions *_requestOptions;
+  id<GRPCResponseHandler> _handler;
+  GRPCCallOptions *_callOptions;
+  Class _responseClass;
+
+  GRPCCall2 *_call;
+  dispatch_queue_t _dispatchQueue;
+}
+
+- (instancetype)initWithRequestOptions:(GRPCRequestOptions *)requestOptions
+                       responseHandler:(id<GRPCResponseHandler>)handler
+                           callOptions:(GRPCCallOptions *)callOptions
+                         responseClass:(Class)responseClass {
+  if ((self = [super init])) {
+    _requestOptions = [requestOptions copy];
+    _handler = handler;
+    _callOptions = [callOptions copy];
+    _responseClass = responseClass;
+    _dispatchQueue = dispatch_queue_create(nil, DISPATCH_QUEUE_SERIAL);
+
+    [self start];
+  }
+  return self;
+}
+
+- (void)start {
+  _call = [[GRPCCall2 alloc] initWithRequestOptions:_requestOptions
+                                    responseHandler:self
+                                        callOptions:_callOptions];
+  [_call start];
+}
+
+- (void)cancel {
+  dispatch_async(_dispatchQueue, ^{
+    if (_call) {
+      [_call cancel];
+      _call = nil;
+    }
+    if (_handler) {
+      id<GRPCResponseHandler> handler = _handler;
+      dispatch_async(handler.dispatchQueue, ^{
+        [handler closedWithTrailingMetadata:nil
+                                      error:[NSError errorWithDomain:kGRPCErrorDomain
+                                                                code:GRPCErrorCodeCancelled
+                                                            userInfo:@{
+                                                              NSLocalizedDescriptionKey :
+                                                                  @"Canceled by app"
+                                                            }]];
+      });
+      _handler = nil;
+    }
+  });
+}
+
+- (void)writeWithMessage:(GPBMessage *)message {
+  if (![message isKindOfClass:[GPBMessage class]]) {
+    [NSException raise:NSInvalidArgumentException format:@"Data must be a valid protobuf type."];
+  }
+
+  dispatch_async(_dispatchQueue, ^{
+    if (_call) {
+      [_call writeData:[message data]];
+    }
+  });
+}
+
+- (void)finish {
+  dispatch_async(_dispatchQueue, ^{
+    if (_call) {
+      [_call finish];
+      _call = nil;
+    }
+  });
+}
+
+- (void)receivedInitialMetadata:(NSDictionary *)initialMetadata {
+  if (_handler) {
+    id<GRPCResponseHandler> handler = _handler;
```

Comment:
IIUC if we use `self->_handler` it actually retains *self* and dereference it inside the block. Since the dispatched block actually runs in the handler's dispatch queue, using `self->_handler` may race with `_handler = nil` running in the call's dispatch queue.

Reply:
That's right. And it's not only for cancellation; server's normal termination is also taken care of.